File: python/vcp_gateway/audit.py
"""Audit events (SPEC §19, §20).

Every invocation MUST emit a signed audit event, OpenTelemetry-compatible.
Audit events MUST NOT contain secrets and SHOULD carry only hashes of sensitive
arguments. Designed to be consumed by a ledger substrate (mcp-ledger).
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Any, Optional

from vcp_sdk.signing import Signer, sign_document

__all__ = ["audit_event", "AuditLog"]


def _now_iso() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def audit_event(
    *,
    event: str,
    subject: str,
    capability_id: str,
    decision: str,
    trace_id: Optional[str] = None,
    span_id: Optional[str] = None,
    host: Optional[str] = None,
    model: Optional[str] = None,
    provider: Optional[str] = None,
    plan_hash: Optional[str] = None,
    argument_hash: Optional[str] = None,
    grant_id: Optional[str] = None,
    reason_code: Optional[str] = None,
    effect: Optional[str] = None,
    result_hash: Optional[str] = None,
    effect_committed: Optional[bool] = None,
    budget_spent: Optional[dict] = None,
    attestation_ref: Optional[dict] = None,
    timestamp: Optional[str] = None,
    signer: Optional[Signer] = None,
) -> dict:
    """Build (and optionally sign) an audit event (audit-event.schema.json).

    Only hashes of sensitive material are carried; no raw secrets or argument
    values (§19). Required fields: event, trace_id, subject, capability_id,
    decision, timestamp.
    """
    ev: dict[str, Any] = {
        "event": event,
        "trace_id": trace_id or uuid.uuid4().hex,
        "subject": subject,
        "capability_id": capability_id,
        "decision": decision,
        "timestamp": timestamp or _now_iso(),
    }
    optional = {
        "span_id": span_id,
        "host": host,
        "model": model,
        "provider": provider,
        "plan_hash": plan_hash,
        "argument_hash": argument_hash,
        "grant_id": grant_id,
        "reason_code": reason_code,
        "effect": effect,
        "result_hash": result_hash,
        "effect_committed": effect_committed,
        "budget_spent": budget_spent,
        "attestation_ref": attestation_ref,
    }
    for k, v in optional.items():
        if v is not None:
            ev[k] = v

    if signer is not None:
        ev = sign_document(ev, signer, signature_field="signature")
    return ev


class AuditLog:
    """An in-memory append-only audit sink (a stand-in for mcp-ledger)."""

    def __init__(self) -> None:
        self.events: list[dict] = []

    def emit(self, event: dict) -> dict:
        self.events.append(event)
        return event

    def __len__(self) -> int:  # pragma: no cover - convenience
        return len(self.events)

File: go/gateway/audit.go
package gateway

import (
	"github.com/hassard0/vcp-servers/go/sdk"
)

// AuditEvent is a signed, OpenTelemetry-compatible record of an invocation
// (spec §20, schemas/audit-event.schema.json). It MUST NOT contain secrets and
// SHOULD carry only hashes of sensitive arguments (spec §19).
type AuditEvent struct {
	Event           string         `json:"event"`
	TraceID         string         `json:"trace_id"`
	SpanID          string         `json:"span_id,omitempty"`
	Subject         string         `json:"subject"`
	Host            string         `json:"host,omitempty"`
	Model           string         `json:"model,omitempty"`
	Provider        string         `json:"provider,omitempty"`
	CapabilityID    string         `json:"capability_id"`
	PlanHash        string         `json:"plan_hash,omitempty"`
	ArgumentHash    string         `json:"argument_hash,omitempty"`
	GrantID         string         `json:"grant_id,omitempty"`
	Decision        string         `json:"decision"`
	ReasonCode      string         `json:"reason_code,omitempty"`
	Effect          string         `json:"effect,omitempty"`
	ResultHash      string         `json:"result_hash,omitempty"`
	EffectCommitted *bool          `json:"effect_committed,omitempty"`
	BudgetSpent     *Budget        `json:"budget_spent,omitempty"`
	Timestamp       string         `json:"timestamp"`
	Signature       *sdk.Signature `json:"signature,omitempty"`
}

// Sign signs the audit event over its canonical form with the signature block
// removed (spec §3 rule 4). Signing audit events makes the trail tamper-evident
// for a ledger substrate (spec §20).
func (e *AuditEvent) Sign(s sdk.Signer) error {
	unsigned := *e
	unsigned.Signature = nil
	mp, err := decodeToMap(unsigned)
	if err != nil {
		return err
	}
	delete(mp, "signature")
	sig, err := sdk.SignValue(s, mp)
	if err != nil {
		return err
	}
	e.Signature = &sig
	return nil
}

// AuditSink consumes emitted audit events (e.g. an mcp-ledger, an OTel exporter,
// or an in-memory recorder for tests).
type AuditSink interface {
	Emit(e AuditEvent)
}

// MemoryAuditSink records events in memory; useful for the end-to-end scenario and
// for tests.
type MemoryAuditSink struct {
	Events []AuditEvent
}

// Emit implements AuditSink.
func (s *MemoryAuditSink) Emit(e AuditEvent) {
	s.Events = append(s.Events, e)
}
